Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp

```diff
@@ -62,28 +76,76 @@ void MemoryViewWidget::UpdateFont()
   Update();
 }
 
-static int GetColumnCount(MemoryViewWidget::Type type)
+static int GetTypeSize(MemoryViewWidget::Type type)
 {
   switch (type)
   {
   case MemoryViewWidget::Type::ASCII:
-  case MemoryViewWidget::Type::U8:
-    return 16;
-  case MemoryViewWidget::Type::U16:
-    return 8;
-  case MemoryViewWidget::Type::U32:
+  case MemoryViewWidget::Type::Hex8:
+  case MemoryViewWidget::Type::Unsigned8:
+  case MemoryViewWidget::Type::Signed8:
+    return 1;
+  case MemoryViewWidget::Type::Unsigned16:
+  case MemoryViewWidget::Type::Signed16:
+  case MemoryViewWidget::Type::Hex16:
+    return 2;
+  case MemoryViewWidget::Type::Hex32:
+  case MemoryViewWidget::Type::Unsigned32:
+  case MemoryViewWidget::Type::Signed32:
   case MemoryViewWidget::Type::Float32:
     return 4;
+  case MemoryViewWidget::Type::Double:
+    return 8;
   default:
-    return 0;
+    return 1;
+  }
+}
+
+static int GetCharacterCount(MemoryViewWidget::Type type)
+{
+  switch (type)
+  {
+  case MemoryViewWidget::Type::ASCII:
+    return 1;
+  case MemoryViewWidget::Type::Hex8:
+    return 2;
+  case MemoryViewWidget::Type::Unsigned8:
+    return 3;
+  case MemoryViewWidget::Type::Hex16:
+  case MemoryViewWidget::Type::Signed8:
+    return 4;
+  case MemoryViewWidget::Type::Unsigned16:
+    return 5;
+  case MemoryViewWidget::Type::Signed16:
+    return 6;
+  case MemoryViewWidget::Type::Hex32:
+    return 8;
+  case MemoryViewWidget::Type::Float32:
+    return 9;
+  case MemoryViewWidget::Type::Double:
+  case MemoryViewWidget::Type::Unsigned32:
+  case MemoryViewWidget::Type::Signed32:
+    return 10;
+  default:
+    return 8;
   }
 }
 
 void MemoryViewWidget::Update()
 {
   clearSelection();
 
-  setColumnCount(2 + GetColumnCount(m_type));
+  // Optional Alignments, always a power of 2.
+  u32 address = m_address;
+
+  if (m_align_fixed->isChecked())
+    address = address & -m_bytes_per_row;
+  else if (m_align_auto->isChecked())
+    address = address & -GetTypeSize(m_type);
```

Comment:
I see what you're doing here, but some compilers are very likely to warn about taking a negative of an unsigned here. You probably want to write this as `address & ~(m_bytes_per_row - 1)` instead. Same for the GetTypeSize() one.

Reply:
Actually, I just remembered we have a `Common::AlignDown()` function, maybe use that.

Reply:
Thanks, I was trying to find a support function